geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/utils/ThreadManagement.scala

```diff
@@ -28,28 +36,120 @@ object ThreadManagement extends LazyLogging {
+    val timeout = math.max(1, scan.timeout.absolute - System.currentTimeMillis())
+    executor.schedule(new QueryKiller(scan), timeout, TimeUnit.MILLISECONDS)
+  }
 
   /**
-    * Trait for classes to be managed for timeouts
-    */
-  trait ManagedQuery extends Closeable {
-    def getTimeout: Long
-    def isClosed: Boolean
-    def debug: String
-  }
+   * Trait for scans that are managed, i.e. tracked and terminated if they exceed a specified timeout
+   *
+   * @tparam T type
+   */
+  trait ManagedScan[T] extends CloseableIterator[T] {
+
+    /**
+     * Scan timeout
+     *
+     * @return
+     */
+    def timeout: Timeout
+
+    /**
+     * Low-level scan to be stopped
+     *
+     * @return
+     */
+    protected def underlying: LowLevelScanner[T]
+
+    // used for log messages
+    protected def typeName: String
+    protected def filter: Option[Filter]
+
+    // we can use a volatile var since we only update the value with a single thread
+    @volatile
+    private var terminated = timeout.absolute <= System.currentTimeMillis()
+
+    private val iter = ExceptionalIterator(if (terminated) { Iterator.empty } else { underlying.iterator })
+    private val cancel = if (terminated) { None } else { Some(ThreadManagement.register(this)) }
+
+    // note: check iter.hasNext first so we get updated terminated flag
+    override def hasNext: Boolean = iter.hasNext || terminated
+
+    override def next(): T = {
+      if (terminated) {
+        val e = new RuntimeException(s"Scan terminated due to timeout of ${timeout.relative}ms")
+        iter.suppressed.foreach(e.addSuppressed)
+        throw e
+      } else {
+        iter.next()
+      }
+    }
+
+    /**
+     * Forcibly terminate the scan
+     */
+    def terminate(): Unit = {
+      terminated = true
+      try {
+        logger.warn(
+          s"Stopping scan on schema '$typeName' with filter '${filterToString(filter)}' " +
+              s"based on timeout of ${timeout.relative}ms")
+        underlying.close()
+      } catch {
+        case NonFatal(e) => logger.warn("Error cancelling scan:", e)
+      }
+    }
+
+    /**
+     * Was the scan terminated due to timeout
+     *
+     * @return
+     */
+    def isTerminated: Boolean = terminated
 
-  private class QueryKiller(query: ManagedQuery) extends Runnable {
-    override def run(): Unit = {
-      if (!query.isClosed) {
-        logger.warn(s"Stopping ${query.debug} based on timeout of ${query.getTimeout}ms")
-        try { query.close() } catch {
-          case NonFatal(e) => logger.warn("Error cancelling query:", e)
-        }
+    override def close(): Unit = {
+      cancel.foreach(_.cancel(false))
+      // if terminated, we've already closed the iterator
+      if (!terminated) {
+        underlying.close()
       }
     }
   }
+
+  /**
+   * Low level scanner that can be closed to terminate a scan
+   *
+   * @tparam T type
+   */
+  trait LowLevelScanner[T] extends Closeable {
```

Comment:
Should it be AutoCloseable?  Meh, I lose track of which one we like.

Reply:
Closeable exends AutoCloseable... the only difference is the type of exception that gets thrown so it doesn't matter to us. I believe there is some code in geotools that doesn't work right with AutoCloseable, so we always use Closeable.